keras_nlp/samplers/contrastive_sampler.py

```diff
@@ -0,0 +1,224 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contrastive Sampler."""
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.compiler.tf2xla.python.xla import dynamic_update_slice
+
+from keras_nlp.api_export import keras_nlp_export
+from keras_nlp.samplers.sampler import Sampler
+from keras_nlp.samplers.sampler import call_args_docstring
+from keras_nlp.utils.python_utils import format_docstring
+
+
+@format_docstring(call_args=call_args_docstring)
+@keras_nlp_export("keras_nlp.samplers.ContrastiveSampler")
+class ContrastiveSampler(Sampler):
+    """Contrastive Sampler class.
+
+    This sampler implements contrastive search algorithm. In short, the sampler
+    chooses the token having the max "score" as the next token. The "score" is
+    a weighted sum between token's probability and max similarity against
+    previous tokens. By using this joint score, contrastive sampler reduces the
+    behavior of duplicating seen tokens.
+
+    Args:
+        k: int, the `k` value of top-k. Next token will be chosen from k tokens.
+        alpha: float, the weight of minus max similarity in joint score
+            computation. The larger the value of `alpha`, the score relies more
+            on the similarity than the token probability.
+        seed: int, defaults to None. The random seed.
+
+    Call Args:
+        {{call_args}}
+
+    Examples:
+    ```python
+    # Use a simple alphabet of lowercase characters to [0, 26).
+    int_lookup = {i: chr(i + ord('a')) for i in range(26)}
+    char_lookup = {v: k for k, v in int_lookup.items()}
+    batch_size, length, vocab_size = 1, 12, len(int_lookup)
+
+    def next(prompt, cache, index):
+        # A uniform distribution over our alphabet.
+        logits = tf.ones((batch_size, vocab_size))
+        return logits, None, cache
```

Comment:
I don't think this will work without a hidden state right?

```
hidden_state = tf.ones((batch_size, 1))
logits = tf.ones((batch_size, vocab_size))
```

Reply:
You will also need to do this update for all the other sampler docstrings! Basically our new "contract" is the next function should always return a hidden state.